### PR TITLE
transaction type Loan disbursement added

### PR DESCRIPTION
**Pull request by nikIntrasoft**
Description of the issue/feature this PR addresses:

Current behavior before PR:
Currently on loan disbursement to savings account , the transaction type is shown as 'deposit'.

Desired behavior after PR is merged:
After merging this PR, the transaction of disbursing the loan to savings account would be shown as 'loan disbursement'.

--

### CODE REVIEW

**Comment by nikIntrasoft** [comment]
Here is the sql required for the modification of m_savings_account_transaction

	ALTER TABLE `m_savings_account_transaction`
		ADD COLUMN `isLoanDisbursement` TINYINT(1) NULL DEFAULT '0' AFTER `is_manual`;

**Comment by tarzan0820** [comment]
@nikIntrasoft what is your skype handle?

**Comment by nikIntrasoft** [comment]
nikpawar89

**Comment by tarzan0820** [comment]
@nikIntrasoft thanks for the new PR... I note you have some reservations about this implementation can you please note them here. 

Also we add the sql required for the modification of m_savings_account_transaction as a migration can't we use a really high number (the community advised this for a similar case)?

**Comment by nikIntrasoft** [comment]
Yes, I think adding high number would be a viable option. However, while merging new updates from community you would have to be aware that you don't override our custom changes.. Also there is a possibility that the out of box migration scripts might fail.

Now coming to this design:
We have have made modifications in some crucial parts of code. We have modified SavingsAccountsTransaction and its table structure in database. You have to be aware of this code change and be cautious during applying updates. The very next PR is going to conflict with this change.
 I think its workable, but be aware of our custom changes.

**Comment by ottoman91** [comment]
@tarzan0820 i have reviewed this and also checked it with francis. the implementation has been done according to your suggestions.

**Comment by tarzan0820** [comment]
@nikIntrasoft can you please submit the migration file for this? use some really high number

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsEnumerations.java [hunk 1]
```diff
@@ -201,6 +201,10 @@ public static SavingsAccountTransactionEnumData transactionType(final SavingsAcc
                 optionData = new SavingsAccountTransactionEnumData(SavingsAccountTransactionType.AMOUNT_RELEASE.getValue().longValue(),
                         SavingsAccountTransactionType.AMOUNT_RELEASE.getCode(), "Release Amount");
                 break; 
+            case LOAN_DISBURSEMENT:
+            	optionData = new SavingsAccountTransactionEnumData(SavingsAccountTransactionType.LOAN_DISBURSEMENT.getValue().longValue(),
+					SavingsAccountTransactionType.LOAN_DISBURSEMENT.getCode(), "Loan Disbursement");
+				break;
```
Comment (tarzan0820) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsEnumerations.java]: I propose that we change this line to

optionData = new SavingsAccountTransactionEnumData(SavingsAccountTransactionType.DEPOSIT.getValue().longValue(),
                        SavingsAccountTransactionType.DEPOSIT.getCode(), "Loan Disbursement");

REason been that some sections of the code use the code and id from SavingsAccountTransactionEnumData and we need this id and code to be consistent to the ones expected from deposits. see - org/apache/fineract/accounting/journalentry/service/JournalEntryReadPlatformServiceImpl.java

**Comment by tarzan0820** [comment]
@nikIntrasoft let us write some unit tests to ensure that loan deposits pass some of the most important tests for deposits?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsEnumerations.java (hunk 1 above)
Comment (nikIntrasoft) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsEnumerations.java]: changed as directed

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsAccountReadPlatformServiceImpl.java [hunk 2]
```diff
@@ -774,7 +774,7 @@ public SavingsAccountTransactionsMapper() {
             sqlBuilder.append("tr.id as transactionId, tr.transaction_type_enum as transactionType, ");
             sqlBuilder.append("tr.transaction_date as transactionDate, tr.amount as transactionAmount,");
             sqlBuilder.append("tr.created_date as submittedOnDate,");
-            sqlBuilder.append("tr.running_balance_derived as runningBalance, tr.is_reversed as reversed,");
+            sqlBuilder.append("tr.running_balance_derived as runningBalance, tr.is_reversed as reversed,tr.isLoanDisbursement as isLoanDisbursement,");
```
Comment (nikIntrasoft) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsAccountReadPlatformServiceImpl.java]: this has been handled

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountDomainServiceJpa.java [hunk 3]
```diff
@@ -142,13 +153,13 @@ public SavingsAccountTransaction handleDeposit(final SavingsAccount account, fin
...
         return handleDeposit(account, fmt, transactionDate, transactionAmount, paymentDetail, isAccountTransfer, isRegularTransaction,
-                savingsAccountTransactionType);
+                savingsAccountTransactionType,false);
     }
 
     private SavingsAccountTransaction handleDeposit(final SavingsAccount account, final DateTimeFormatter fmt,
             final LocalDate transactionDate, final BigDecimal transactionAmount, final PaymentDetail paymentDetail,
             final boolean isAccountTransfer, final boolean isRegularTransaction,
-            final SavingsAccountTransactionType savingsAccountTransactionType) {
+            final SavingsAccountTransactionType savingsAccountTransactionType,final boolean isLoanDisbursement) {
```
Comment (nikIntrasoft) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountDomainServiceJpa.java]: done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountDomainServiceJpa.java [hunk 4]
```diff
@@ -197,7 +210,7 @@ public SavingsAccountTransaction handleDividendPayout(final SavingsAccount accou
         final boolean isRegularTransaction = true;
         final SavingsAccountTransactionType savingsAccountTransactionType = SavingsAccountTransactionType.DIVIDEND_PAYOUT;
         return handleDeposit(account, fmt, transactionDate, transactionAmount, paymentDetail, isAccountTransfer, isRegularTransaction,
-                savingsAccountTransactionType);
+                savingsAccountTransactionType,false);
```
Comment (nikIntrasoft) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountDomainServiceJpa.java]: done

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountTransaction.java [hunk 5]
```diff
@@ -117,6 +117,9 @@
     
     @Column(name = "is_manual", length = 1, nullable = true)
     private boolean isManualTransaction;
+    
+    @Column(name = "isLoanDisbursement", length = 1, nullable = true)
```
Comment (nikIntrasoft) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/domain/SavingsAccountTransaction.java]: done

**Pull request merged**